### PR TITLE
Refactor FIPS BootstrapChecks to simple checks

### LINKED ISSUE #34772: FIPS checks should not be bootstrap checks
We currently have FIPS bootstrap checks, but really these do not fit as a bootstrap check since they are always enforced. We should refactor these checks into a `FIPSChecks` class that can be instantiated on startup. cc @jasontedor

### PR DESCRIPTION
**Pull request by jkakavas**
FIPS 140 bootstrap checks should not be bootstrap checks as they
are always enforced. This commit moves the validation logic within
the security plugin.
The FIPS140SecureSettingsBootstrapCheck was not applicable as the
keystore was being loaded on init, before the Bootstrap checks
were checked, so an elasticsearch keystore of version < 3 would
cause the node to fail in a FIPS 140 JVM before the bootstrap check
kicked in, and as such hasn't been migrated.

Resolves: #34772
Replaces: #36677

### CODE REVIEW

**Review by jasontedor — APPROVED** [review]
LGTM.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 1]
```diff
@@ -830,6 +833,41 @@ static void validateRealmSettings(Settings settings) {
...
+        final List<String> validationErrors = new ArrayList<>();
+        Settings keystoreTypeSettings = settings.filter(k -> k.endsWith("keystore.type"))
+            .filter(k -> settings.get(k).equalsIgnoreCase("jks"));
+        if (keystoreTypeSettings.isEmpty() == false) {
+            validationErrors.add("JKS Keystores cannot be used in a FIPS 140 compliant JVM. Please " +
+                "revisit [" + keystoreTypeSettings.toDelimitedString(',') + "] settings");
+        }
+        Settings keystorePathSettings = settings.filter(k -> k.endsWith("keystore.path"))
+            .filter(k -> settings.hasValue(k.replace(".path", ".type")) == false);
+        // Default Keystore type is JKS in JDK8 and PKCS12 in > JDK9 if not explicitly set
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: This is not the case. We explicitly set the keystore type to JKS, and don't depend on the JVM default.
https://github.com/elastic/elasticsearch/blob/9945d5cf35183263184e518aa73a72959fb15a0e/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLConfigurationSettings.java#L245-L252

Perhaps we should call `inferKeyStoreType` here instead.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 2]
```diff
@@ -303,19 +307,15 @@ public Security(Settings settings, final Path configPath) {
         this.env = new Environment(settings, configPath);
         this.enabled = XPackSettings.SECURITY_ENABLED.get(settings);
         if (enabled) {
-            runStartupChecks(settings);
+            runStartupChecks(settings, getLicenseState());
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: Is license state meaningful at this point during node start up?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 3]
```diff
@@ -830,6 +833,41 @@ static void validateRealmSettings(Settings settings) {
...
+                "revisit [" + keystorePathSettings.toDelimitedString(',') + "] settings");
+        }
+        final String selectedAlgorithm = XPackSettings.PASSWORD_HASHING_ALGORITHM.get(settings);
+        if (selectedAlgorithm.toLowerCase(Locale.ROOT).startsWith("pbkdf2") == false) {
+            validationErrors.add("Only PBKDF2 is allowed for password hashing in a FIPS 140 JVM. Please set the " +
+                "appropriate value for [ " + XPackSettings.PASSWORD_HASHING_ALGORITHM.getKey() + " ] setting.");
+        }
+
+        if (licenseState != null && FIPS_ALLOWED_LICENSE_OPERATION_MODES.contains(licenseState.getOperationMode()) == false) {
+            validationErrors.add("FIPS mode is only allowed with a Platinum or Trial license");
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: The error message should be built from `FIPS_ALLOWED_LICENSE_OPERATION_MODES`

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 4]
```diff
@@ -273,6 +275,8 @@
         DiscoveryPlugin, MapperPlugin, ExtensiblePlugin {
 
     private static final Logger logger = LogManager.getLogger(Security.class);
+    static final EnumSet<License.OperationMode> FIPS_ALLOWED_LICENSE_OPERATION_MODES =
+        EnumSet.of(License.OperationMode.PLATINUM, License.OperationMode.TRIAL);
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: Can we move this to `XPackLicenseState`? We ought to keep licensing decisions centralised.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java [hunk 5]
```diff
@@ -253,7 +255,7 @@ public void testJoinValidatorForFIPSLicense() throws Exception {
         new Security.ValidateLicenseForFIPS(false).accept(node, state);
 
         final boolean isLicenseValidForFips =
-            FIPS140LicenseBootstrapCheck.ALLOWED_LICENSE_OPERATION_MODES.contains(license.operationMode());
+            Security.FIPS_ALLOWED_LICENSE_OPERATION_MODES.contains(license.operationMode());
```
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java]: This (existing) test is testing 2 different scenarios based on randomness in the license type.
Can we move it to 2 explicit test cases?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 6]
```diff
@@ -326,8 +326,11 @@ public Security(Settings settings, final Path configPath) {
 
     }
 
-    private static void runStartupChecks(Settings settings) {
```
[the remark this answers is not in the document]
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: Leftover change from uncompleted attempt I'd guess, wasn't left on purpose

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java (hunk 2 above)
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: I see your point. Should we just remove checks for license on node startup and only leave `ValidateLicenseForFIPS` on node joins ?

**Comment by jkakavas** [comment]
@tvernum I believe I have addressed your comments, if you want to take another look at this

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Comment by jkakavas** [comment]
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request+multijob+packaging-tests-unix-next-sample/os=ubuntu-16.04-packaging/415/console

> 22:57:21 FAILURE: Build failed with an exception.
> 22:57:21 
> 22:57:21 * What went wrong:
> 22:57:21 Execution failed for task ':distribution:docker:buildDockerImage'.

```
22:44:27 
22:44:27 
22:44:27  One of the configured repositories failed (Unknown),
22:44:27  and yum doesn't have enough cached data to continue. At this point the only
22:44:27  safe thing yum can do is fail. There are a few ways to work "fix" this:
22:44:27 
22:44:27      1. Contact the upstream for the repository and get them to fix the problem.
22:44:27 
22:44:27      2. Reconfigure the baseurl/etc. for the repository, to point to a working
22:44:27         upstream. This is most often useful if you are using a newer
22:44:27         distribution release than is supported by the repository (and the
22:44:27         packages for the previous distribution release still work).
22:44:27 
22:44:27      3. Run the command with the repository temporarily disabled
22:44:27             yum --disablerepo=<repoid> ...
22:44:27 
22:44:27      4. Disable the repository permanently, so yum won't use it by default. Yum
22:44:27         will then just ignore the repository until you permanently enable it
22:44:27         again or use --enablerepo for temporary usage:
22:44:27 
22:44:27             yum-config-manager --disable <repoid>
22:44:27         or
22:44:27             subscription-manager repos --disable=<repoid>
22:44:27 
22:44:27      5. Configure the failing repository to be skipped, if it is unavailable.
22:44:27         Note that yum will try to contact the repo. when it runs most commands,
22:44:27         so will have to try and fail each time (and thus. yum will be be much
22:44:27         slower). If it is a very temporary problem though, this is often a nice
22:44:27         compromise:
22:44:27 
22:44:27             yum-config-manager --save --setopt=<repoid>.skip_if_unavailable=true
22:44:27 
22:44:27 Cannot find a valid baseurl for repo: base/7/x86_64
22:44:27 Could not retrieve mirrorlist http://mirrorlist.centos.org/?release=7&arch=x86_64&repo=os&infra=container error was
22:44:27 14: curl#6 - "Could not resolve host: mirrorlist.centos.org; Unknown error"
```

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Review by tvernum — APPROVED** [review]

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Comment by jkakavas** [comment]
@elasticmachine update branch just in case this will ever work :)

**Comment by jkakavas** [comment]
@elasticmachine update branch

**Comment by jkakavas** [comment]
@elasticmachine elasticsearch-ci/packaging-sample-matrix

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Pull request merged**